File: server/ruby/app/infrastructure/mappers/additional_payment_actions/capture_payment_mapper.rb
module Infrastructure
  module Mappers
    module AdditionalPaymentActions
      module CapturePaymentMapper
        def self.to_sdk_request(dto)
          request = OnlinePayments::SDK::Domain::CapturePaymentRequest.new
          request.amount = dto.amount
          request.is_final = dto.is_final
          request
        end
        def self.from_sdk_response(response)
          response_dto = Business::Dtos::AdditionalPaymentActions::ResponseDto.new

          response_dto.status =
            response&.status ? Business::Domain::Enums::AdditionalPaymentActions::Status.try_from(response.status) : nil

          response_dto.id = response&.id

          status_output = Business::Domain::Payments::StatusOutput.new
          status_output.status_code = response&.status_output&.status_code
          status_output.status_category = nil

          response_dto.status_output = status_output

          response_dto
        end
      end
    end
  end
end

File: server/python/app/infrastructure/mappers/additional_payment_actions/capture_payment_mapper.py
from onlinepayments.sdk.domain.capture_payment_request import CapturePaymentRequest
from app.application.domain.enums.additional_payment_actions.status import AdditionalPaymentActionStatus
from app.application.domain.payments.status_output import StatusOutput
from app.application.dtos.additional_payment_actions.request_dto import AdditionalPaymentActionsRequestDto
from app.application.dtos.additional_payment_actions.response_dto import AdditionalPaymentActionsResponseDto

class CapturePaymentMapper:

    @staticmethod
    def to_sdk_request(dto: AdditionalPaymentActionsRequestDto) -> CapturePaymentRequest:
        request = CapturePaymentRequest()
        request.amount = dto.amount
        request.is_final = dto.is_final

        return request

    @staticmethod
    def from_sdk_response(response) -> AdditionalPaymentActionsResponseDto:
        response_dto = AdditionalPaymentActionsResponseDto()

        response_dto.id = getattr(response, 'id', None)

        status = getattr(response, 'status', None)
        response_dto.status = AdditionalPaymentActionStatus.try_from(status) if status else None

        status_output = StatusOutput()
        status_output_obj = getattr(response, 'status_output', None)
        status_output.status_code = getattr(status_output_obj, 'status_code', None)
        status_output.status_category = None

        response_dto.status_output = status_output

        return response_dto